sdk/storage/Azure.Storage.Common/tests/Shared/RecordedClientRequestIdPolicy.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Azure.Core.Http;
+using Azure.Core.Pipeline;
+using Azure.Core.Testing;
+
+namespace Azure.Storage.Test.Shared
+{
+    /// <summary>
+    /// Pipeline policy to verify x-ms-client-request-id and x-ms-client-return-request-id 
+    /// headers that are echoed back from a request match.
+    /// </summary>
+    public class RecordedClientRequestIdPolicy : SynchronousHttpPipelinePolicy
+    {
+        private readonly TestRecording _testRecording;
+        private readonly string _parallelRangePrefix = null;
+
+        /// <summary>
+        /// Create a new RecordedClientRequestIdPolicy
+        /// </summary>
+        public RecordedClientRequestIdPolicy(TestRecording testRecording, bool parallelRange = false)
+        {
+            _testRecording = testRecording;
+            if (parallelRange)
+            {
+                _parallelRangePrefix = _testRecording.GenerateId() + "_";
+            }
+        }
+
+        /// <summary>
+        /// Verify x-ms-client-request-id and x-ms-client-return-request-id headers matches as
+        /// x-ms-client-return-request-id is an echo of x-mis-client-request-id.
+        /// </summary>
+        /// <param name="message">The message that was sent</param>
+        public override void OnSendingRequest(HttpPipelineMessage message)
+        {
+            if (_parallelRangePrefix != null &&
+                message.Request.Headers.TryGetValue("x-ms-range", out string range))
+            {
+                // If we're transferring a sequence of ranges in parallel, use
+                // the same prefix and use the range to differentiate each message
+                message.Request.ClientRequestId = _parallelRangePrefix + range;
+            }
+            else if(_parallelRangePrefix != null &&
+                message.Request.UriBuilder.Query.Contains("blockid"))
+            {
+                var queryParameters = message.Request.UriBuilder.Query.Split('&');
```

Comment:
You can have UriQueryParamsCollection in the Common project to do this parsing for you

Reply:
We're going to leave this for now, it appears this change would cause us to have to re-record tests again, and we've already spent enough time on this.

Reply:
You don't have to make this fix now, but you also wouldn't have to re-record the tests.

Reply:
There was almost definitely something else going on, but when we tried to implement this it was only working with re-recorded tests.